keps/sig-api-machinery/1623-standardize-conditions/README.md

```diff
@@ -0,0 +1,144 @@
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+- [ ] Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] KEP approvers have approved the KEP status as `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+While many Kuberentes APIs have `.status.conditions`, the schema of `condition` varies a lot between them.
+There is very little commonality at the level of serialization, proto-encoding, and required vs optional.
+Conditions are central enough to the API to make a common golang type with a fixed schema.
+The schema can be a strong recommendation to all API authors.
+
+## Motivation
+
+Allow general consumers to expect a common schema for `.status.conditions` and share golang logic for common Get, Set, Is for `.status.conditions`.
+The pattern is well-established and we have a good sense of the schema we now want.
+
+### Goals
+
+ 1. For all new APIs, have a common type for `.status.conditions`.
+ 2. Provide common utility methods for `HasCondition`, `IsConditionTrue`, `SetCondition`, etc.
+ 3. Provide recommended defaulting functions that set required fields and can be embedded into conversion/default functions.
+
+### Non-Goals
+
+ 1. Update all existing APIs to make use of the new condition type.
+
+## Proposal
+
+Introduce a type into k8s.io/apimachinery/pkg/apis/meta/v1 for `Condition` that looks like 
+```go
+type Condition struct {
+	// Type of condition in CamelCase.
+	// +required
+	Type string `json:"type" protobuf:"bytes,1,opt,name=type"`
+	// Status of the condition, one of True, False, Unknown.
+	// +required
+	Status ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status"`
+	// Last time the condition transitioned from one status to another.
+    // This should be when the underlying condition changed.  If that is not known, then using the time when the API field changed is acceptable.
+	// +required
+	LastTransitionTime metav1.Time `json:"lastTransitionTime" protobuf:"bytes,3,opt,name=lastTransitionTime"`
```

Comment:
How about using metav1.MicroTime instead? This sometimes make performance tracking easier.

Reply:
> How about using metav1.MicroTime instead? This sometimes make performance tracking easier.

I wouldn't generally expect conditions to be used to try to track performance.

Reply:
I definitely wouldn't add something on purpose to track performance, but given we already have this field, we can get something for free by insignificant cost of changing to MicroTime. So why not?

Reply:
> I definitely wouldn't add something on purpose to track performance, but given we already have this field, we can get something for free by insignificant cost of changing to MicroTime. So why not?

Because that doesn't match any known .status.conditions schema.  This KEP isn't inventing a new thing.  It's standardizing the existing thing.

Reply:
Not sure I agree with the argument, given you wrote that what you propose doesn't match any existing schema anyway:
`This is not strictly compatible with any of our existing conditions because of either proto ordinals,
required vs optional, or omitEmpty or not."

The changes from Time to MicroTime are mechanical, and can be done with sed...

Reply:
> Not sure I agree with the argument, given you wrote that what you propose doesn't match any existing schema anyway:
> `This is not strictly compatible with any of our existing conditions because of either proto ordinals,
> required vs optional, or omitEmpty or not."
> 
> The changes from Time to MicroTime are mechanical, and can be done with sed...

It's about the consumers, not the implementers.  By being compatible with what exists today, we can make it easier for consumers to adopt.  If I saw a big gain to be had by making the change I would be in favor, but I don't think that the gain warrants the pain to consumers that are used to "normal" time.

Reply:
OK - I see what you're saying now. That makes sense to me, given that I don't have any immediate usecase for it.